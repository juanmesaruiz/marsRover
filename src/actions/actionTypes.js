export const ROVER_CLEAR = 'ROVER@CLEAR';
export const ROVER_CLEAR_POSITION_LOG = 'ROVER@CLEAR_POSITION_LOG';
export const ROVER_DELETE_OBSTACLE = 'ROVER@DELETE_OBSTACLE';
export const ROVER_NEW_INSTRUCTIONS = 'ROVER@NEW_INSTRUCTIONS';
export const ROVER_NEW_MOVE = 'ROVER@NEW_MOVE';
export const ROVER_SET_GRID = 'ROVER@SET_GRID';
export const ROVER_SET_OBSTACLES = 'ROVER@SET_OBSTACLES';
export const ROVER_SET_POSITION = 'ROVER@SET_POSITION';
