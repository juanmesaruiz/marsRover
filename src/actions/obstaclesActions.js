import { ROVER_SET_OBSTACLES } from './actionTypes';

export const createObstacles = obstacles => ({
  type: ROVER_SET_OBSTACLES,
  payload: {
    obstacles,
  },
});
