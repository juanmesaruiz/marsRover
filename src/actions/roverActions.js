import { ROVER_CLEAR, ROVER_NEW_MOVE, ROVER_SET_POSITION } from './actionTypes';

export const roverClear = () => ({
  type: ROVER_CLEAR,
});

export const roverSetPosition = position => ({
  type: ROVER_SET_POSITION,
  payload: {
    position,
  },
});

export const roverNewKeyboardMove = ({ code, grid, obstaclesCoordinates }) => ({
  type: ROVER_NEW_MOVE,
  payload: {
    code,
    grid,
    obstaclesCoordinates,
  },
});
