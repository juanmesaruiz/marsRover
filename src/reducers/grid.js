import { ROVER_CLEAR } from '../actions/actionTypes';

const defaultState = { x: 50, y: 25 };

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    default:
      return state;
  }
};

export default grid;
