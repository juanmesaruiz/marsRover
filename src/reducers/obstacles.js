import { ROVER_CLEAR, ROVER_SET_OBSTACLES } from '../actions/actionTypes';

const defaultState = [];

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    case ROVER_SET_OBSTACLES: {
      const { obstacles } = action.payload;
      return [...state, ...obstacles];
    }

    default:
      return state;
  }
};

export default grid;
