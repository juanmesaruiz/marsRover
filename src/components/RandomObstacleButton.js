import React, { useCallback } from 'react';
import { connect } from 'react-redux';

import { createObstacles } from '../actions/obstaclesActions';
import { getGrid, getObstacles, getRover } from '../reducers';

import Button from './library/Button';

import { getNewObstaclesCoordinates } from '../common/helpers';

const RandomObstacleButton = ({
  createObstacles,
  grid,
  obstaclesCoordinates,
  rover: { current },
}) => {
  const handleCreateRandomObstacle = useCallback(
    () =>
      createObstacles(
        getNewObstaclesCoordinates(grid, current, obstaclesCoordinates)
      ),
    [createObstacles, current, grid, obstaclesCoordinates]
  );

  return (
    <Button onClick={handleCreateRandomObstacle}>Add random obstacle</Button>
  );
};

const mapStateToProps = state => ({
  grid: getGrid(state),
  obstaclesCoordinates: getObstacles(state),
  rover: getRover(state),
});

export default connect(mapStateToProps, { createObstacles })(
  RandomObstacleButton
);
