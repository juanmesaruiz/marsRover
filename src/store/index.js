import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import reducers from '../reducers';

const configureStore = () => {
  const initialState = {};
  const enhancers = [];
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension({ name: '- ROVER MARS -' }));
  }

  const composedEnhancers = compose(applyMiddleware(thunk), ...enhancers);

  const store = createStore(reducers, initialState, composedEnhancers);

  return { store };
};

export default configureStore;
