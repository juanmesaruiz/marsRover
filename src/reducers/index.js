import { combineReducers } from 'redux';

import grid from './grid';
import obstacles from './obstacles';
import rover from './rover';

const reducers = {
  grid,
  obstacles,
  rover,
};

export const getGrid = state => state.grid;
export const getObstacles = state => state.obstacles;
export const getRover = state => state.rover;

export default combineReducers(reducers);
