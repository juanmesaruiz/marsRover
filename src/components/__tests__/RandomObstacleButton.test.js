import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { createObstacles } from '../../actions/obstaclesActions';
import { getGrid, getObstacles, getRover } from '../../reducers';

import RandomObstacleButton from '../RandomObstacleButton';

import state from '../../store/__mocks__/mockState';
import store from '../../store/__mocks__/mockStore';

jest.mock('../../actions/obstaclesActions');
jest.mock('../../reducers');

describe('RandomObstacleButton test suite', () => {
  const component = (
    <Provider store={store}>
      <RandomObstacleButton />
    </Provider>
  );

  const setup = () => render(component);

  beforeEach(() => {
    createObstacles.mockReturnValue({
      type: 'createObstacles',
    });
    getGrid.mockReturnValue(state.grid);
    getObstacles.mockReturnValue(state.obstacles);
    getRover.mockReturnValue(state.rover);
  });

  afterEach(jest.clearAllMocks);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should fire createObstacles action', () => {
    const { getByText } = setup();
    fireEvent.click(getByText('Add random obstacle'));

    expect(createObstacles).toHaveBeenCalledTimes(1);
  });
});
