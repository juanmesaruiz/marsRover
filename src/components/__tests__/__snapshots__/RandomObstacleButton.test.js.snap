// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomObstacleButton test suite Should match snapshot 1`] = `
.c0 {
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    Add random obstacle
  </button>
</div>
`;
